### PR TITLE
Projection/Backprojection Adjointedness

### PR DESCRIPTION
**Pull request by garrettwrong**
See #1109 .

The following hard coded unit tests would need to be addressed.  They are the ones I would expect, no surprises yet.  They are all using the `LegacyVolume`/`_LegacySimulation` and I have no issue refactoring them.

I'm also wondering if this can resolve the spurious transpose in `estimate`.

```
================================================= short test summary info ==================================================
FAILED tests/test_mean_estimator.py::MeanEstimatorTestCase::testAdjoint - NameError: name 'Simulation' is not defined
FAILED tests/test_mean_estimator.py::MeanEstimatorTestCase::testEstimate - NameError: name 'Simulation' is not defined
FAILED tests/test_mean_estimator_boosting.py::test_fsc[volume=<class 'aspire.volume.volume_synthesis.AsymmetricVolume'>, order=None-resolution=32-dtype=<class 'numpy.float32'>] - AssertionError: 
FAILED tests/test_mean_estimator_boosting.py::test_mse[volume=<class 'aspire.volume.volume_synthesis.AsymmetricVolume'>, order=None-resolution=32-dtype=<class 'numpy.float32'>] - AssertionError: 
FAILED tests/test_mean_estimator_boosting.py::test_fsc[volume=<class 'aspire.volume.volume_synthesis.CnSymmetricVolume'>, order=4-resolution=32-dtype=<class 'numpy.float32'>] - AssertionError: 
FAILED tests/test_mean_estimator_boosting.py::test_mse[volume=<class 'aspire.volume.volume_synthesis.CnSymmetricVolume'>, order=4-resolution=32-dtype=<class 'numpy.float32'>] - AssertionError: 
FAILED tests/test_mean_estimator_boosting.py::test_fsc[volume=<class 'aspire.volume.volume_synthesis.CnSymmetricVolume'>, order=5-resolution=32-dtype=<class 'numpy.float32'>] - AssertionError: 
FAILED tests/test_mean_estimator_boosting.py::test_mse[volume=<class 'aspire.volume.volume_synthesis.CnSymmetricVolume'>, order=5-resolution=32-dtype=<class 'numpy.float32'>] - AssertionError: 
FAILED tests/test_mean_estimator_boosting.py::test_fsc[volume=<class 'aspire.volume.volume_synthesis.DnSymmetricVolume'>, order=2-resolution=32-dtype=<class 'numpy.float32'>] - AssertionError: 
FAILED tests/test_mean_estimator_boosting.py::test_mse[volume=<class 'aspire.volume.volume_synthesis.DnSymmetricVolume'>, order=2-resolution=32-dtype=<class 'numpy.float32'>] - AssertionError: 
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! KeyboardInterrupt !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
/Users/gbwright/opt/miniconda3/envs/imgprjadj/lib/python3.8/site-packages/finufft/_interfaces.py:287: KeyboardInterrupt
(to show a full traceback on KeyboardInterrupt use --full-trace)
============================================ 10 failed, 7 deselected in 36.25s =============================================

```

### CODE REVIEW

**Comment by garrettwrong** [comment]
Need to replace the `optimize` tests, which were checking a hardcoded call to conj_grad.

I will migrate the mean est and weighted mean est test files towards fixtures while here, since not much is left of the original tests now anyway...

**Comment by garrettwrong** [comment]
After considering the `optimize` tests more, I was at a loss how to replace them that would not just be a hardcode mock, and what value that actually brought, so I just culled them.  The (conj_grad) code path is tested in any other tests using mean volume estimation.

Otherwise the mean estimator and weighted volume estimator test files were migrated toward pytest fixtures and then extended to cover the usual product of dtype, even-odd pixels, and the two different preconditioner options.

I think the last things left are to resolve this patch for cov3d and self review.

**Review by j-c-c — DISMISSED** [review]
Looks good!

File: tests/test_mean_estimator.py [hunk 1]
```diff
@@ -1,733 +1,166 @@
...
+    mean_b_coef = estimator.src_backward()
+
+    # Evaluate mean coefs into a volume
+    est = Coef(basis, mean_b_coef).evaluate() * mask
+
+    # Mask off corners of volume
+    vol = sim.vols * mask
+
+    # Assert the mean volume is close to original volume
+    np.testing.assert_allclose(
```
Comment (janden) [inline tests/test_mean_estimator.py]: Why should this work? The `mean_b_coeff` contains the right-hand side coefficients of the normal equation. To get an estimate of the mean volume, you need to invert the kernel first.

File: tests/test_mean_estimator.py [hunk 2]
```diff
@@ -1,733 +1,166 @@
...
-                _ = estimator.estimate()
-
-            # Load the checkpoint coefficients while tmp_input_dir exists.
-            b_chk = np.load(f"{prefix}_iter{test_iter:04d}.npy")
+        # Assert we raise when reading `maxiter`.
+        with raises(RuntimeError, match="Unable to converge!"):
+            _ = _estimator.estimate()
+
+        # Load the checkpoint coefficients while tmp_input_dir exists.
+        b_chk = np.load(f"{prefix}_iter{test_iter:04d}.npy")
```
Comment (janden) [inline tests/test_mean_estimator.py]: Is it the b (RHS) coefficients that are stored or the current estimate coefficients?

File: tests/test_weighted_mean_estimator.py [hunk 3]
```diff
@@ -1,787 +1,187 @@
...
+    # Construct simple test weights;
+    # one uniform positive and negative weighted volume respectively.
+    r = 2  # Number of weighted volumes
+    weights = np.ones((sim.n, r)) / np.sqrt(sim.n)
+    weights[:, 1] *= -1  # negate second weight vector
+
+    return weights
+
+
+@pytest.fixture(
```
Comment (janden) [inline tests/test_weighted_mean_estimator.py:80]: Shouldn't this define a `preconditioner` fixture?

**Review by janden — CHANGES_REQUESTED** [review]
Looks pretty good. Some minor questions.

On a more general note, should we not have an explicit test that the projection and backprojections are actually adjoints of one another (similar to the type of test that was reported in the issue)?

File: tests/test_weighted_mean_estimator.py (hunk 3 above)
Comment (garrettwrong) [reply tests/test_weighted_mean_estimator.py:80]: ~None is the default "circulant".  `"None"` is no preconditioner .  This is confusing and the codes should probably be rewritten, but I was only trying to fix the reported bug.~

Take that back. You are correct, `None` was not defaulting to circulant.  Updated to test all three cases.

File: tests/test_mean_estimator.py (hunk 1 above)
Comment (garrettwrong) [reply tests/test_mean_estimator.py]: I thought this was an more/less an unfiltered backprojection and should be similar to the original volume.  Is that not the right way to think about it?

It does capture the effect of the bug.

File: tests/test_mean_estimator.py (hunk 2 above)
Comment (garrettwrong) [reply tests/test_mean_estimator.py]: Ah, oops, poor name on my part.  The checkpoint saves are the `x` estimate from the last iteration.  I'll change to `x` something, thanks!

Comment (garrettwrong) [reply tests/test_mean_estimator.py]: Passing through to `x0`, renamed `x_chk`. My mistake.

Comment (janden) [reply tests/test_mean_estimator.py]: Thanks!

File: tests/test_weighted_mean_estimator.py (hunk 3 above)
Comment (janden) [reply tests/test_weighted_mean_estimator.py:80]: Great. Thanks!

**Comment by garrettwrong** [comment]
I've added a direct test of the adjoint relationship to `test_mean_estimator`.  Going to attempt adding a test of src_backward now (previously incorrectly called test_adjoint).

**Comment by garrettwrong** [comment]
I was able to add a test using the `src_backward` method. It required scaling by `n` images.  I will look into that further and check the test with a few more situations.

**Comment by garrettwrong** [comment]
OKay, added src_backward test.  Looks like probably need to live with the scaling factor.  Not straightfoward way to scale `backproject` because it only know about the image batch is handling (and they may be different sizes...).

I added a commit ([27d89b4](https://github.com/ComputationalCryoEM/ASPIRE-Python/pull/1110/commits/27d89b4bd4d50bb6f98f859e0869b1ce2af380ab)) that changes the interface of `im_backward` in a way that I _think_ is safer.  However, that change is totally optional and I'm fine with backing it out.  What I was concerned with regarding the current implementation is that is allows for potentially mismatches Image and index relationships, unless I misread it.  Can discuss.

File: src/aspire/source/image.py [hunk 4]
```diff
@@ -937,31 +937,38 @@ def normalize_background(self, bg_radius=1.0, do_ramp=True):
             LambdaXform(normalize_bg, bg_radius=bg_radius, do_ramp=do_ramp)
         )
 
-    def im_backward(self, im, start, weights=None, symmetry_group=None):
+    def im_backward(self, idx, im=None, weights=None, symmetry_group=None):
```
Comment (garrettwrong) [inline src/aspire/source/image.py]: This is the change I mention in the PR comments.  Can roll it back if the prior interface is more clear to others.

**Comment by garrettwrong** [comment]
Closes #1109

File: tests/test_mean_estimator.py (hunk 1 above)
Comment (garrettwrong) [reply tests/test_mean_estimator.py]: discussed in team meeting, replaced test.

**Review by j-c-c — DISMISSED** [review]
Looks good to me.

**Comment by garrettwrong** [comment]
@janden , reverted the interface change discussed this morning.  PR should cover your prior suggestions and be ready for review. Thanks

**Review by janden — APPROVED** [review]
Looks great. Thank1

**Pull request merged**